palace/linalg/errorestimator.cpp

```diff
@@ -0,0 +1,332 @@
+  pcg->SetRelTol(tol);
+  pcg->SetAbsTol(std::numeric_limits<double>::epsilon());
+  pcg->SetMaxIter(max_it);
+
+  ksp = std::make_unique<KspSolver>(std::move(pcg), std::move(pc));
+  ksp->SetOperators(*M, *M);
+}
+
+CurlFluxErrorEstimator::CurlFluxErrorEstimator(
+    const IoData &iodata, const MaterialOperator &mat_op,
+    mfem::ParFiniteElementSpaceHierarchy &nd_fespaces)
+  : mat_op(mat_op), nd_fespaces(nd_fespaces),
+    smooth_projector(nd_fespaces, iodata.solver.linear.tol, 200, 1,
+                     iodata.solver.pa_order_threshold),
+    smooth_flux(nd_fespaces.GetFinestFESpace().GetTrueVSize()),
+    flux_rhs(nd_fespaces.GetFinestFESpace().GetTrueVSize()),
+    field_gf(&nd_fespaces.GetFinestFESpace()),
+    smooth_flux_gf(&nd_fespaces.GetFinestFESpace())
+{
+}
+
+template <>
+ErrorIndicators CurlFluxErrorEstimator::ComputeIndicators(const ComplexVector &v) const
+{
+  auto &nd_fespace = nd_fespaces.GetFinestFESpace();
+  const int nelem = nd_fespace.GetNE();
+
+  Vector smooth_vec, coarse_vec, estimates(nelem);
+  estimates = 0.0;
+  double normalization = 0.0;
+  for (bool real : {true, false})
+  {
+    field_gf.SetFromTrueDofs(real ? v.Real() : v.Imag());
+
+    // Coefficients for computing the discontinuous flux component, i.e. (W, μ⁻¹∇ × V).
+    CurlFluxCoefficient coef(field_gf, mat_op);
+    {
+      // Given the RHS vector of non-smooth flux, construct a flux projector and perform
+      // mass matrix inversion in the appropriate space, giving f = M⁻¹ f̂.
+      Mpi::Print("Computing smooth flux approximation of {} component\n",
+                 real ? "real" : "imaginary");
+      BlockTimer bt(Timer::ESTSOLVE);
+      mfem::LinearForm rhs(&nd_fespace);
+      rhs.AddDomainIntegrator(new VectorFEDomainLFIntegrator(coef));
+      rhs.UseFastAssembly(false);
+      rhs.Assemble();
+      nd_fespace.GetProlongationMatrix()->MultTranspose(rhs, flux_rhs);
+      smooth_projector.Mult(flux_rhs, smooth_flux);
+    }
+    smooth_flux_gf.SetFromTrueDofs(smooth_flux);
+    smooth_flux_gf.ExchangeFaceNbrData();
+
+    // Loop over elements and accumulate the estimates from this component
+    for (int e = 0; e < nd_fespace.GetNE(); e++)
+    {
+      auto &T = *nd_fespace.GetElementTransformation(e);
+      // integration order 2p + q
+      const auto &ir = mfem::IntRules.Get(T.GetGeometryType(),
+                                          2 * nd_fespace.GetFE(e)->GetOrder() + T.Order());
+      for (const auto &ip : ir)
+      {
+        T.SetIntPoint(&ip);
+
+        smooth_flux_gf.GetVectorValue(e, ip, smooth_vec);
+        coef.Eval(coarse_vec, T, ip);
+        double w_i = ip.weight * T.Weight();
+        constexpr int vdim = 3;
+        for (int c = 0; c < vdim; c++)
+        {
+          estimates[e] += w_i * std::pow(smooth_vec[c] - coarse_vec[c], 2.0);
+          normalization += w_i * coarse_vec[c] * coarse_vec[c];
+        }
+      }
+    }
+  }
+  for (auto &e : estimates)
+  {
+    e = std::sqrt(e);
```

Comment:
Can move into the loop above?

Reply:
Each element is visited once for each component, accumulating the square, so this has to be done afterwards.